crates/bitwarden-ssh/src/import.rs

```diff
@@ -0,0 +1,214 @@
+use ed25519;
+use pkcs8::{der::Decode, pkcs5, DecodePrivateKey, PrivateKeyInfo, SecretDocument};
+use ssh_key::private::{Ed25519Keypair, RsaKeypair};
+
+use crate::{error::SshKeyImportError, SshKey};
+
+const PKCS1_LABEL: &str = "RSA PRIVATE KEY";
+const PKCS8_UNENCRYPTED_LABEL: &str = "PRIVATE KEY";
+const PKCS8_ENCRYPTED_LABEL: &str = "ENCRYPTED PRIVATE KEY";
+const OPENSSH_LABEL: &str = "OPENSSH PRIVATE KEY";
+
+pub fn import_key(
+    encoded_key: String,
+    password: Option<String>,
+) -> Result<SshKey, SshKeyImportError> {
+    let label = pem_rfc7468::decode_label(encoded_key.as_bytes())
+        .map_err(|_| SshKeyImportError::ParsingError)?;
+
+    match label {
+        PKCS1_LABEL => Err(SshKeyImportError::UnsupportedKeyType),
+        PKCS8_UNENCRYPTED_LABEL => import_pkcs8_key(encoded_key, None),
+        PKCS8_ENCRYPTED_LABEL => import_pkcs8_key(
+            encoded_key,
+            Some(password.ok_or(SshKeyImportError::PasswordRequired)?),
+        ),
+        OPENSSH_LABEL => import_openssh_key(encoded_key, password),
+        _ => Err(SshKeyImportError::ParsingError),
+    }
+}
+
+fn import_pkcs8_key(
+    encoded_key: String,
+    password: Option<String>,
+) -> Result<SshKey, SshKeyImportError> {
+    let der = if let Some(password) = password {
+        SecretDocument::from_pkcs8_encrypted_pem(&encoded_key, password.as_bytes()).map_err(
+            |err| match err {
+                pkcs8::Error::EncryptedPrivateKey(pkcs5::Error::DecryptFailed) => {
+                    SshKeyImportError::WrongPassword
+                }
+                _ => SshKeyImportError::ParsingError,
+            },
+        )?
+    } else {
+        SecretDocument::from_pkcs8_pem(&encoded_key).map_err(|_| SshKeyImportError::ParsingError)?
+    };
+
+    let private_key_info =
+        PrivateKeyInfo::from_der(der.as_bytes()).map_err(|_| SshKeyImportError::ParsingError)?;
+
+    let private_key = match private_key_info.algorithm.oid {
+        ed25519::pkcs8::ALGORITHM_OID => {
+            let private_key: ed25519::KeypairBytes = private_key_info
+                .try_into()
+                .map_err(|_| SshKeyImportError::ParsingError)?;
+            ssh_key::private::PrivateKey::from(Ed25519Keypair::from(&private_key.secret_key.into()))
```

Comment:
🤔 Three type conversions on a single line leads me to believe that there should be a more direct way to convert key formats. It looks like all of these types come from a library? If so, consider wrapping the logic in a [newtype to implement an external trait on an external type](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types).